(feat) basic rename

dummdidumm opened the pull request:
#110

- [x] rename where everything is inside same svelte file works
- [x] rename of variable inside svelte which is also in another ts/js file works
- [ ] rename of svelte prop works when doing the rename for prop of component A in component B which uses that prop
- [x] rename of svelte prop works when doing the rename for prop of component A in component A

Out of scope for this PR:
-  rename in ts file which would affect svelte file -> out of this scope because we would also need to trigger the language server on ts/js files then

jasonlyu123 commented:
Found an issue. if I rename a props that is html attribute like disabled. It would do a rename on the html attribute no matter that props exist on the component or not.

dummdidumm commented:
Good catch, thank you. There is also a prop rename problem. I think I need to enhance `svelte2tsx` with some diagnostic capabilities like "what other svelte components does this file import?" or "what props does this file have?".

dummdidumm commented:
I ended up using `getRenameInfo` of the ts language service. The downside is that component properties can only be renamed within that component now. I added a TODO about this, this can be tackled later. I also added `prepareRename` support so it shows the user early that this cannot be renamed.

dummdidumm commented:
@jasonlyu123 could you take one more look? 😃 Also code-wise, I had to jump through some hoops to get it working.

jasonlyu123 commented on packages/language-server/src/lib/documents/utils.ts:
```diff
@@ -215,3 +215,33 @@ export function isInTag(position: Position, tagInfo: TagInformation | null): boo
+    return getVariableAtOffset(offsetAt(position, text), text);
+}
+
+export function getVariableAtOffset(offset: number, text: string) {
```
also variable.b, variable) and operators without space in between.

jasonlyu123 replied:
I found a hack that can be used to find the identifier/symbol in the generated tsx:
```ts
const offset = fragment.offsetAt(fragment.getGeneratedPosition(position));
const { start, length } = lang.getSmartSelectionRange(tsDoc.filePath, offset).textSpan;
const identifier = tsDoc.getText(start, start + length);
```

jasonlyu123 commented:
```svelte
<input bind:value />
```
would be renamed to
```svelte
<input bind:valuvalues />
```

jasonlyu123 commented on packages/language-server/src/plugins/typescript/features/RenameProvider.ts:
```diff
@@ -0,0 +1,325 @@
+        // First find out if it's really the "rename prop inside component with that prop" case
+        // Use original document for that because only there the `export` is present.
+        const regex = new RegExp(
+            `export\\s+(const|let)\\s+${getVariableAtPosition(
```
@sw-yx do you know about export const, function, class? I only found it to be available in the component instance, if so it probably doesn't needs to be checked.

dummdidumm commented:
This is a source map problem, seems like `svelte2tsx` mappings are off after conversion. I was able to fix it. I also added your suggestion for the variable retrieval and added a TODO that `export {x as y}` syntax is currently not supported for rename. The function/class thing ... let's just wait till someone shows up with this request 😄
